File: src/2770_Maximum_Number_of_Jumps_to_Reach_the_Last_Index/solution.py
## https://leetcode.com/problems/maximum-number-of-jumps-to-reach-the-last-index/

from typing import List

## TC: O(N^2)
## SC: O(N)
def maximumJumps(nums: List[int], target: int) -> int:
    n = len(nums)
    dp = [-1] * n
    dp[0] = 0
    for i in range(1, n):
        for j in range(i - 1, -1, -1):
            if abs(nums[i] - nums[j]) <= target:
                if dp[j] > -1:
                    dp[i] = max(dp[i], dp[j] + 1)
    return dp[-1]

File: src/2770_Maximum_Number_of_Jumps_to_Reach_the_Last_Index/solution.ts
// https://leetcode.com/problems/maximum-number-of-jumps-to-reach-the-last-index/

// TC: O(N^2)
// SC: O(N)
function maximumJumps(nums: number[], target: number): number {
  const n = nums.length;
  let dp = new Array<number>(n).fill(-1);
  dp[0] = 0;

  for (let i = 1; i < n; i++) {
    for (let j = i - 1; j > -1; j--) {
      if (Math.abs(nums[i] - nums[j]) <= target) {
        if (dp[j] > -1) {
          dp[i] = Math.max(dp[i], dp[j] + 1);
        }
      }
    }
  }
  return dp[n - 1];
}
